engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/TypesFromSignaturesTest.java

```diff
@@ -0,0 +1,186 @@
+                """
+                    type A
+                    type B
+                    type C
+                        Value v
+
+                    f0 = 0
+
+                    f4 x y z w = [x, y, z, w]
+
+                    f2 : A -> B -> C
+                    f2 x y = [x, y]
+                    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = compile(src);
+
+    var f0 = findStaticMethod(module, "f0");
+    var f4 = findStaticMethod(module, "f4");
+    var f2 = findStaticMethod(module, "f2");
+
+    // For 0 arguments and unknown return type we know nothing useful, so no information is
+    // registered.
+    assertNoInferredType(f0);
+
+    // For a function without ascriptions, we can at least infer the _arity_
+    // Currently that is denoted by replacing unknowns with Any. Later this may be free type
+    // variables.
+    assertInferredType(f4, "Any -> (Any -> (Any -> (Any -> Any)))");
+
+    // For the 'opted-out' ascription, the types are ignored, because they are not checked types.
+    // But we still infer arity.
+    assertInferredType(f2, "Any -> (Any -> Any)");
+  }
+
+  @Test
+  public void memberMethods() throws URISyntaxException {
+    final URI uri = new URI("memory://memberMethods.enso");
+    final Source src =
+        Source.newBuilder(
+                "enso",
+                """
+                    type A
+                        static_method (x : A) -> A = x
+                        member_method self (x : A) -> A = x
+                    standalone_method (x : A) -> A = x
+                    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = compile(src);
+    var staticMethod = findMemberMethod(module, "A", "static_method");
+    var memberMethod = findMemberMethod(module, "A", "member_method");
+
+    assertInferredType(staticMethod, "A -> A");
+    assertInferredType(memberMethod, "A -> A");
+  }
+
+  @Test
+  public void extensionMethods() throws URISyntaxException {
+    final URI uri = new URI("memory://extensionMethods.enso");
+    final Source src =
+        Source.newBuilder(
+                "enso",
+                """
+                    type A
+
+                    A.extension_static_method (x : A) -> A = x
+                    A.extension_member_method self (x : A) -> A = x
+                    """,
+                uri.getAuthority())
+            .uri(uri)
+            .buildLiteral();
+
+    var module = compile(src);
+    var staticMethod = findMemberMethod(module, "A", "extension_static_method");
+    var memberMethod = findMemberMethod(module, "A", "extension_member_method");
+
+    assertInferredType(staticMethod, "A -> A");
+    assertInferredType(memberMethod, "A -> A");
+  }
+
+  private void assertInferredType(IR ir, String expected) {
+    TypeRepresentation inferred = getInferredType(ir);
+    assertEquals(expected, inferred.toString());
```

Comment:
Minor: Relying on correct `toString` implementation in tests does not seem like a good idea. What about moving the current `toString` implementation to a class in tests dir in the same package? It would then look something like this:
```java
assertEquals(expected, TestUtil.toString(inferred));
```

Reply:
> Relying on correct `toString` implementation in tests does not seem like a good idea.

Hmm, I never looked at it this way. IMHO it may make sense to test `toString`, but I guess it's a habit from working on Libs where the `to_text` implementation is part of the public user-facing API and is _supposed_ to be well tested.

I'm happy to move it to a helper if you think that will be better.

Reply:
> Relying on correct `toString` implementation in tests does not seem like a good idea

Why? `toString()` format is part of an API. But yeah, having for example:
```
assertTypeRepresentation(....)
```
might be more flexible.